File: ocd-elixir/lib/ocd.ex
defmodule Ocd do
  import Ecto.Query
  alias Ocd.Repo
  alias Ocd.Db.AuditLog

  @type csv_row :: {integer(), DateTime.t()}

  def main do
    records = parse_csv_file()
    outcrop_ids = Enum.map(records, fn {id, _} -> id end)
    IO.puts("Parsed #{length(records)} records from CSV file")

    logs = existing_audit_logs(outcrop_ids)
    IO.puts("Loaded #{length(logs)} existing audit logs")

    to_insert = logs_to_insert(records, logs)
    IO.puts("#{length(to_insert)} audit logs need to be created")

    if length(to_insert) > 0 do
      {success, failed} = create_audit_logs(to_insert)
      IO.puts("Inserted #{success} logs successfully and #{failed} errors")
    else
      IO.puts("Nothing to do!")
    end
  end

  @spec parse_csv_file() :: [csv_row()]
  defp parse_csv_file do
    File.stream!("dates.csv")
    |> CSV.decode(separator: ?;, headers: true)
    |> Stream.map(&parse_csv_row/1)
    |> Enum.to_list()
  end

  @spec parse_csv_row({:ok, map()} | any()) :: csv_row()
  defp parse_csv_row({:ok, %{"ID" => id, "Inserted At" => inserted_at}}) do
    {id, _} = Integer.parse(id)
    inserted_at = Timex.parse!(inserted_at, "%Y-%m-%d %H:%M:%S", :strftime)
    {id, inserted_at}
  end

  defp parse_csv_row(_), do: raise("Error parsing CSV row")

  @spec existing_audit_logs([integer()]) :: [%AuditLog{}]
  defp existing_audit_logs(ids) do
    from(al in AuditLog,
      where: al.outcrop_id in ^ids and al.action == "created" and al.entity == "outcrop"
    )
    |> Repo.all()
  end

  @spec logs_to_insert([csv_row()], [%AuditLog{}]) :: [csv_row()]
  defp logs_to_insert(csv_data, existing_logs) do
    existing_ids = Enum.map(existing_logs, fn log -> log.outcrop_id end)
    Enum.filter(csv_data, fn {id, _} -> !Enum.member?(existing_ids, id) end)
  end

  @spec create_audit_logs([csv_row()]) :: {num_success :: integer(), num_failed :: integer()}
  defp create_audit_logs(items) when length(items) > 0 do
    items =
      Enum.map(items, fn {outcrop_id, inserted_at} ->
        %{
          entity: "outcrop",
          action: "created",
          user_id: 11,
          outcrop_id: outcrop_id,
          study_id: nil,
          inserted_at: inserted_at
        }
      end)

    {num_success, _} = Repo.insert_all(AuditLog, items)
    {num_success, length(items) - num_success}
  end
end

File: OCD.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}

module OCD (backfill) where

import Control.Monad.IO.Class
import Control.Monad.Logger (NoLoggingT (runNoLoggingT))
import qualified Data.ByteString.Lazy as BL
import Data.Csv
import Data.Functor ((<&>))
import Data.Int
import Data.Maybe
import Data.Text (Text)
import Data.Time (UTCTime, defaultTimeLocale, parseTimeM)
import qualified Data.Vector as V
import Database.Persist.Postgresql
import Database.Persist.TH

data OcInfo = OcInfo {outcropId :: Int64, insertedAt :: UTCTime} deriving (Show)

instance FromNamedRecord OcInfo where
  parseNamedRecord m =
    OcInfo
      <$> m .: "ID"
      <*> (m .: "Inserted At" >>= parseDate)
    where
      parseDate :: String -> Parser UTCTime
      parseDate = parseTimeM True defaultTimeLocale "%Y-%m-%d %T"

parseCsvFile :: String -> IO (Either String [OcInfo])
parseCsvFile filePath = do
  csvData <- BL.readFile filePath
  case decodeByNameWith decodeOpts csvData :: Either String (Header, V.Vector OcInfo) of
    Left err -> pure $ Left err
    Right (_, v) -> pure $ Right (V.toList v)
  where
    decodeOpts = defaultDecodeOptions {decDelimiter = fromIntegral (fromEnum ';')}

share
  [mkPersist sqlSettings]
  [persistLowerCase|
AuditLog sql=audit_logs
    entity Text
    action Text
    userId Int64
    outcropId Int64 Maybe
    studyId Int64 Maybe
    insertedAt UTCTime
    deriving Show
|]

execQuery :: SqlPersistT IO a -> IO a
execQuery action = runNoLoggingT $
  withPostgresqlConn "postgres://postgres:postgres@localhost:5432/safari_api" $
    \backend -> liftIO $ runSqlConn action backend

ocIdsWithExistingLogs :: [Int64] -> IO [Int64]
ocIdsWithExistingLogs outcropIds = execQuery $ do
  selectList
    [AuditLogEntity ==. "outcrop", AuditLogOutcropId <-. map Just outcropIds]
    []
    <&> mapMaybe (auditLogOutcropId . entityVal)

insertNewLogs :: [(Int64, UTCTime)] -> IO (Int, Int)
insertNewLogs toBeCreated = do
  results <- execQuery $ insertMany $ map mapLog toBeCreated
  let numSuccess = length $ filter ((/=) $ toSqlKey 0) results
  let numFailed = length results - numSuccess
  pure (numSuccess, numFailed)
  where
    mapLog (ocId, inserted) =
      AuditLog
        { auditLogEntity = "outcrop",
          auditLogAction = "created",
          auditLogUserId = 11,
          auditLogOutcropId = Just ocId,
          auditLogStudyId = Nothing,
          auditLogInsertedAt = inserted
        }

backfill :: IO ()
backfill = do
  parseResult <- parseCsvFile "dates.csv"
  case parseResult of
    Left err -> putStrLn $ "Error parsing CSV:" ++ err
    Right ocInfos -> do
      ocIds <- ocIdsWithExistingLogs $ map outcropId ocInfos
      putStrLn $ "Found " ++ show (length ocIds) ++ " existing OCs with logs"
      let toInsert = filter (\ocInfo -> outcropId ocInfo `notElem` ocIds) ocInfos
      putStrLn $ show (length toInsert) ++ " records to be created"
      (success, failed) <- insertNewLogs $ map (\ocInfo -> (outcropId ocInfo, insertedAt ocInfo)) toInsert
      putStrLn $ "Created " ++ show success ++ " successful logs with " ++ show failed ++ " failed."